### DISCUSSION TITLE
[turborepo]  Usage of Recoil in Nextjs app induces a TypeError: TypeError: Cannot destructure property 'ReactCurrentDispatcher' of '__TURBOPACK__imported__

### DISCUSSION THREAD

**Discussion by HayatoKazi** [vercel/turborepo#8373] [Help]
### Summary

I have added store as package in the turborepo, which contains Recoil as state management. A hook named "useBalance" was added, which returns the state of an atom.(basic unit of recoil). Exported the hook file. Added the "store" package as dependency to the Next.js app. 
The issue arises whenever I try to invoke "useBalance" hook in the Next.js app. 

Steps taken to solve the issue:
1. Updated Node and Npm versions to the latest.
2. removed node_modules of turbo repo and removed .next folder of Next.js App. Reinstalled again.
3. Changed the node version value in engine of package.json of the root .

Steps to reproduce the issue:
https://github.com/tsdineshjai/PayTM_Repo

Navigate to apps/user-app and npm run dev. Click localhost:3000

Key Issue: Integration issue of Recoil into NextJs app

![image](https://github.com/vercel/turbo/assets/59303778/138ea617-dd28-40e0-a240-f038b0bb0c0f)
![image](https://github.com/vercel/turbo/assets/59303778/019662c6-f789-4a70-8ce7-e625e6026f49)

### Additional information

```js
"recoil": "^0.7.7"
"next": "15.0.0-rc.0",
"node" : "v22.2.0",
"npm" : "10.8.1"
```

### Example

https://github.com/tsdineshjai/PayTM_Repo

**Comment by HayatoKazi** [comment]
Do I need to create an issue for this discussion or this will be suffice ?

**Reply by anthonyshew** [reply]
To make sure I understand, do you believe this to be an issue with Turborepo or Turbopack? From what I'm seeing here, this appears to be for Turbopack?

**Reply by HayatoKazi** [reply]
> To make sure I understand, do you believe this to be an issue with Turborepo or Turbopack? From what I'm seeing here, this appears to be for Turbopack?

I have installed the TurboRepo via this link, https://turbo.build/repo/docs/getting-started. I believe that this issue is related to Turborepo.

**Reply by anthonyshew** [reply]
Turborepo [doesn't play a part in dependency resolution](https://turbo.build/repo/docs/crafting-your-repository/managing-dependencies#turborepo-does-not-manage-dependencies) is why I'm asking. Do you see the same behavior when you're using Recoil with Turbopack in a standalone application (not a monorepo)?

**Reply by HayatoKazi** [reply]
Yeah, I tried to do so in a Turbopack standalone app, I didn't see any issues. It is perfectly fine. :)

Would you be able to tell me where the issue might be?

**Reply by anthonyshew** [reply]
Interesting. Can you open an issue on the Next.js repo with your reproduction? I'm suspecting that this is a Turbopack bug.

https://github.com/vercel/next.js/issues/new?assignees=&labels=bug&projects=&template=1.bug_report.yml

**Reply by HayatoKazi** [reply]
> Interesting. Can you open an issue on the Next.js repo with your reproduction? I'm suspecting that this is a Turbopack bug.
> 
> https://github.com/vercel/next.js/issues/new?assignees=&labels=bug&projects=&template=1.bug_report.yml

As suggested, I have opened an issue on Next. Js repo. Should I tag you in that issue ? 
https://github.com/vercel/next.js/issues/66695

**Reply by Yashita1512** [reply]
I was facing the same issue, finally found what was wrong. 
So here are a few steps that may help you :

1. Remove node_modules, turbo cache(.turbo folder) and package-lock.json file from the root folder. (for uninstalling all dependencies and clearing the cached files).
2. Add "recoil" as a dependency in both, web/doc's package.json and packages/store's (the atoms store's) package.json
3. Then again install dependencies in the root folder. 

Turbo will now install recoil in the root folder node_modules alongwith all other dependencies/libraries and make it available for those that added it as a dependency.
As turborepo does build system orchestration we need to specify the relevant dependencies for each of the workspaces as and when used. (Workspaces as in packages and apps(mentioning for clarity)).

**Reply by HayatoKazi** [reply]
> ackages and apps

Thanks. I will give it a try and will respond to you if it works.

**Reply by HayatoKazi** [reply]
> > ackages and apps
> 
> Thanks. I will give it a try and will respond to you if it works.

It didnt work for me.

**Reply by Lakshyyaa** [reply]
hey going through the exact same problem, any luck?
i just cant install recoil in a next app in turborepo and if i use --force i feel i'll eventually end up messing up

**Reply by Yashita1512** [reply]
One update - Also try deleting the .next and node_modules folders from both the apps. Apparently the .next cache is causing the type error and not allowing next to use the newly installed dependencies. I faced the same issue with the 'docs' app and nothing was working, finally this worked for me.

**Reply by Yashita1512** [reply]
This was another step to be done before trying the above. I posted it 2 days back, I think it was missed.

Here's my package.json code for the next-app, store and root turborepo. Replace these and try installing recoil in the root folder. It should work.

https://opposite-wax-112.notion.site/package-json-for-all-3-9f68dd2d5e204ce783f24c5c319ef515

**Comment by Yashita1512** [comment]
Here's my package.json code for the next-app, store and root turborepo. Replace these and try installing recoil in the root folder. It should work.

https://opposite-wax-112.notion.site/package-json-for-all-3-9f68dd2d5e204ce783f24c5c319ef515

**Reply by Rishiii7** [reply]
Thanks! it's working for me

**Comment by faraaz-inc** [comment]
After banging my head on this issue for hours, now i think the problem lies with Recoil not supporting React 19 RC that came out last month(check out this issue in recoil's repo https://github.com/facebookexperimental/Recoil/issues/2318 ). 
Thats why we are facing this issue as create-turbo@latest uses this version currently.
@Yashita1512 is having react 18, thats why it works.

**Reply by Rishiii7** [reply]
Yeah, when i was installing recoil in user-app, I could not install it giving me an npm error. Afterwords I changed the react, react-dom version to 18.3.1 and next to ^12.0.0 . Still didn't work for me.   

It woked as mentioned by @Yashita1512

**Comment by Rethick11** [comment]
"next": "^14.1.1",
"react": "^18.2.0",
"react-dom": "^18.2.0",
 "recoil": "^0.7.7"
    
    It only works with above configuration in next app and 
    change next.config.ts => next.config.js and in that file paste


    /** @type {import('next').NextConfig} */
module.exports = {
  
};
issue solved for me :)

**Reply by udayempire** [reply]
I installed react 18 but stil facing the issue

**Comment by ArshdeepSingh728281** [comment]
People
recoil does not support ssr so do not use it in next js
for redux you need to make changes in layout.tsx
so use **zustand (ssr friendly)** easy syntax.

**Reply by SarthkK** [reply]
agreed, recoil has been archived since 1 jan 2025.
Switch to zustand or jotai.
